extractor/main.py

```diff
@@ -23,78 +25,164 @@
 )
 from pathlib import Path
 
-# Load the data into a DataFrame
-# climate_action_library_test.csv for testing and changing values
-# climate_action_library_original.csv for original C40 list
-df = load_datafile_into_df(Path("../data/climate_action_library_test.csv"))
-
-# Prepare a list to hold all mapped data
-mapped_data = []
-
-# Define a list of extraction functions and corresponding JSON fields
-extraction_functions = {
-    # "ActionID": extract_ActionID, -> not in the mapping because it is already set in the main script as incremental index
-    "ActionName": extract_ActionName,
-    # "ActionType": extract_ActionType, -> not in the mapping because it is already extracted in the main script as first step
-    "AdaptationCategory": extract_AdaptationCategory,
-    "Hazard": extract_Hazard,
-    # "Sector": extract_Sector,
-    "Subsector": extract_Subsector,
-    "PrimaryPurpose": extract_PrimaryPurpose,
-    "InterventionType": extract_InterventionType,
-    "Description": extract_Description,
-    "BehavioralChangeTargeted": extract_BehavioralChangeTargeted,
-    "CoBenefits": extract_CoBenefits,
-    "EquityAndInclusionConsiderations": extract_EquityAndInclusionConsiderations,
-    "GHGReductionPotential": extract_GHGReductionPotential,
-    "AdaptionEffectiveness": extract_AdaptionEffectiveness,
-    "CostInvestmentNeeded": extract_CostInvestmentNeeded,
-    "TimelineForImplementation": extract_TimelineForImplementation,
-    "Dependencies": extract_Dependencies,
-    "KeyPerformanceIndicators": extract_KeyPerformanceIndicators,
-    "Impacts": extract_Impacts,
-}
-
-# For testing, only process x rows
-df_subset = df  # .head(2)
-
-# Incremental counter for ActionID
-action_id = 1
-
-# Iterate over DataFrame rows
-for index, df_row in df_subset.iterrows():
+# Create a semaphore with a concurrency limit (e.g., 10 tasks at a time)
+semaphore = asyncio.Semaphore(100)
```

Comment:
Nice i didn't see semaphores used in this way clever thing i might still

Reply:
Yea its a great way of not running into any sorts of API limits for LLM or for more compute intensive scripts into resource limits